modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala

```diff
@@ -48,5 +59,67 @@ object ConstructionService {
 
       NonEmptyList.fromList(accountIdentifiers)
     }
+
+    def transactionToOperations(transaction: Transaction): Either[String, NonEmptyList[Operation]] =
+      refineV[Not[Equal[0L]]](transaction.amount.value.value).flatMap { amount =>
+        val operations: Either[String, List[Operation]] = refineV[Not[Equal[0L]]](-1L * transaction.amount.value.value).map { negAmount =>
+          val negativeOperation = Operation(
+            OperationIdentifier(OperationIndex(0L)),
+            none,
+            OperationType.Transfer,
+            none,
+            AccountIdentifier(transaction.destination, none).some,
+            Amount(AmountValue(negAmount), DAG).some
+          )
+
+          val positiveOperation = Operation(
+            OperationIdentifier(OperationIndex(1L)),
+            none,
+            OperationType.Transfer,
+            none,
+            AccountIdentifier(transaction.destination, none).some,
+            Amount(AmountValue(amount), DAG).some
+          )
+
+          List(negativeOperation, positiveOperation)
+        }
+
+        operations.flatMap { operationList =>
+          NonEmptyList.fromList(operationList) match {
+            case Some(value) => value.asRight
+            case None        => "No valid operations found.".asLeft
+          }
+        }
+      }
+
+    def parseSignedTransaction(hex: Hex): F[Either[ConstructionError, ConstructionParse.ServiceReturn]] =
+      KryoSerializer[F]
+        .deserialize[Signed[Transaction]](hex.toBytes)
+        .traverse { signedTransaction =>
+          transactionToOperations(signedTransaction).traverse { operationsList =>
+            signedTransaction.proofs.toNonEmptyList.traverse(signature => signature.id.hex.toPublicKey).map { proofs =>
+              val proofAccountIdentifiers = proofs.map(_.toAddress).map { address =>
+                AccountIdentifier(address, none)
+              }
+
+              ConstructionParse.ServiceReturn(operationsList, proofAccountIdentifiers.some)
+            }
+          }.map(_.left.map(_ => MalformedTransaction))
+        }
+        .map(_.left.map(_ => MalformedTransaction).flatten)
```

Comment:
`map` and then `flatten` is just a `flatMap`.

Also if you have `map` inside `map` then it means that you are missing `flatMap` somewhere because these methods look almost the same. See my other comments.

Reply:
I switched to `for-comprehension`